orm/index.go

```diff
@@ -183,33 +216,61 @@ func (i Index) move(db weave.KVStore, prev Object, save Object) error {
 		return ErrModifiedPK()
 	}
 
-	// if the keys don't change, then
-	oldKey, err := i.index(prev)
+	oldKeys, err := i.index(prev)
 	if err != nil {
 		return err
 	}
-	newKey, err := i.index(save)
+	newKeys, err := i.index(save)
 	if err != nil {
 		return err
 	}
-	if bytes.Equal(oldKey, newKey) {
-		return nil
+
+	oldKeysSet := make(map[string]struct{})
+	for _, v := range oldKeys {
+		oldKeysSet[string(v)] = struct{}{}
 	}
 
-	// check unique constraint before removing
-	if i.unique && len(oldKey) != 0 {
-		k := i.IndexKey(newKey)
-		val := db.Get(k)
-		if val != nil {
-			return ErrUniqueConstraint(i.name)
+	newKeysSet := make(map[string]struct{})
+	for _, v := range newKeys {
+		newKeysSet[string(v)] = struct{}{}
+	}
+
+	// check unique constraints first
+	for _, newKey := range newKeys {
+		if _, exists := oldKeysSet[string(newKey)]; exists { // skip existing keys
+			continue
+		}
+		// check unique constraint before removing
+		if i.unique && len(oldKeysSet) != 0 {
+			k := i.IndexKey(newKey)
+			val := db.Get(k)
+			if val != nil {
+				return ErrUniqueConstraint(i.name)
+			}
 		}
 	}
 
-	err = i.remove(db, oldKey, prev.Key())
-	if err != nil {
-		return err
+	// remove unused keys
+	for _, oldKey := range oldKeys {
+		if _, exists := newKeysSet[string(oldKey)]; !exists {
+			if err = i.remove(db, oldKey, prev.Key()); err != nil {
+				return err
+			}
+			continue
+		}
+	}
+
+	// add new keys
+	for _, newKey := range newKeys {
+		// check unique constraint before removing
```

Comment:
`s/removing/inserting/`

also above (my typo in original)

Reply:
the constraint is checked on the insert function as well. So it makes sense only as you wrote it before to check before removing. Although I assume this is happening within a transaction.